janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java

```diff
@@ -241,7 +262,30 @@ public static boolean isEmpty(Condition<?> condition) {
     private static <E extends JanusGraphElement> Or<E> addConstraint(final RelationType type, OrJanusPredicate predicate, List<Object> values, Or<E> or, StandardJanusGraphTx tx) {
         for (int i = 0 ; i < values.size(); i++) {
             final JanusGraphPredicate janusGraphPredicate = predicate.get(i);
-            if (janusGraphPredicate instanceof AndJanusPredicate) {
+            if (janusGraphPredicate instanceof Contain) {
+                //Rewrite contains conditions
+                final Collection childValues = (Collection) values.get(i);
+                if (janusGraphPredicate == Contain.NOT_IN) {
+                    if (childValues.size() == 1) {
+                        addConstraint(type, Cmp.NOT_EQUAL, childValues.iterator().next(), or, tx);
+                    }
+                    // Don't need to handle the case where childValues is empty, because it defaults to
+                    // an or(and()) is added, which is a tautology
+                    final And<E> nested = new And<>(childValues.size());
+                    for (final Object inValue : childValues) {
+                        addConstraint(type, Cmp.NOT_EQUAL, inValue, nested, tx);
+                    }
+                    or.add(nested);
+                } else {
+                    Preconditions.checkArgument(janusGraphPredicate == Contain.IN);
+                    if (childValues.isEmpty()) {
+                        continue; // Handle unsatisfiable condition within or like it does not exist
```

Comment:
nitpick: it took me a while to understand the comment. I guess you mean "Handle unsatisfiable condition `or(within())` like it does not exist"

Reply:
It was actually meant to say "Handle any unsatisfiable condition within an OR statement like it does not exist."